streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java

```diff
@@ -83,6 +96,104 @@ public void closeStore() {
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis()));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis()));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis()));
+        driver.advanceWallClockTime(10 * 1000L);
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 10 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 10 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 10 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 10 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 10 * 1000L));
+        driver.advanceWallClockTime(10 * 1000L);
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 20 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 20 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 20 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 20 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 20 * 1000L));
+        driver.advanceWallClockTime(10 * 1000L);
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 30 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 30 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 30 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 30 * 1000L));
+        driver.pipeInput(recordFactory.create(topic, UUID.randomUUID().toString(), UUID.randomUUID().toString(), System.currentTimeMillis() + 30 * 1000L));
```

Comment:
Is it intended that all records fall into a single window / segment? What was the root cause of the bug and how does this data/timestamp pattern cover it? (cannot remember)

Reply:
It is actually not the root cause: as long as we have records with somewhat different timestamps, they may trigger the issue in caching layer for range queries.